api/v1alpha3/common_types.go

```diff
@@ -59,6 +59,8 @@ const (
 	MachineInternalIP  MachineAddressType = "InternalIP"
 	MachineExternalDNS MachineAddressType = "ExternalDNS"
 	MachineInternalDNS MachineAddressType = "InternalDNS"
+
+	MachineNodeNameIndex = "status.nodeRef.name"
```

Comment:
I think I understand why you moved the const here, but also I find kind of strange to get this constant - which defines something internal to the controllers - in the versioned API

A possible solution is to create a new package named index or something similar under controllers and move under this package
- The index creation (to be called from main)
- The indexMachineByNodeName 
- A new func that gets the machine from a node using the index (this code is now duplicated in machine controller/nodeToMachine and MHC controller/getMachineFromNode)

wdyt?

Reply:
Hmm, while this might seem out of place, I do agree that it might need to be versioned. That said, (apologies in advance) but now I think I'm a little more inclined to have the index in the Machine controller itself

Reply:
I agree on keeping the indexer in Machine controller, it will be more contained this way as in it will keep us from creating it for envtest too.

Reply:
kk
I still think that a new package could provide a cleaner solution, but I don't want to block on this.

Reply:
Open an issue, it's a good idea — someone from the community could pick up this cleanup work in the future after we brainstorm a bit more about it

Reply:
I don't necessarily want to hold up the PR given the discussion about resolving the location in a follow-up, but a few things aren't sitting right with me about the current situation:

1. This `MachineNodeNameIndex` constant isn't anything that would ever be used outside of setting up the index, so it's weird to make this part of an exported public API
1. An index on a field is API version specific, so I agree that it makes sense to think about a field index in the scope of a specific API version package
1. Field indices are global and not per controller, so it feels a bit strange to configure an index inside a controller...
    1. ... unless you need to use information from the controller, but ...
1. ... This particular index does not use anything outside of the Machine struct

Would it be worth considering moving all field index registrations into the appropriate versioned API package(s), where the only thing we expose is an exported method named something like `RegisterFieldIndexFunctions()`, and all the inner workings can be left private (e.g. we wouldn't need to export `MachineNodeNameIndex`). We'd also need to remember to invoke the registration function from `main.go`. This works as long as our index functions are simple and don't require access to a client or something from a controller (although if we need a client, we could set up a way to pass one in to the index function).

Thoughts?

Reply:
For this particular case, we will only need to pass the manager as argument to RegisterFieldIndexFunctions(). 
Then we will call this in envtest as well.
I think we can do this change in this PR given that milestone is 0.4.0 release.

Reply:
Actually we are using `MachineNodeNameIndex ` in nodeToMachine() mapping function in Machine and MachineHealthCheck controllers. I guess we can move the mapper function to the api too?

Reply:
For the time being, I would create a 2nd constant in the controller package that the controller can use.

Reply:
> Would it be worth considering moving all field index registrations into the appropriate versioned API package(s), where the only thing we expose is an exported method named something like RegisterFieldIndexFunctions(), and all the inner workings can be left private (e.g. we wouldn't need to export MachineNodeNameIndex). We'd also need to remember to invoke the registration function from main.go. This works as long as our index functions are simple and don't require access to a client or something from a controller (although if we need a client, we could set up a way to pass one in to the index function).

This makes sense to me, it could also live in the API types itself, just a different file that folks could call as `clusterv1.AddIndexToManager()` or with the function like your suggestion. We can open a different issue to tackle this separately though

Reply:
Tracking this here: https://github.com/kubernetes-sigs/cluster-api/issues/3778